Add Announcement API

AnuraagReddy123 opened the pull request:
# Description

Added `Announcements.kt` and `Reaction.kt` entities and added `AnnouncementMethods.kt` method so far. Need to get these steps reviewed
Closes #300 

# Type of Change

(Keep the one that applies, remove the rest)

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested using
`./gradleew check`

# Mandatory Checklist

- [x] My change follows the projects coding style
- [x] I ran `gradle check` and there were no errors reported
- [x] I have performed a self-review of my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Optional Things To Check

- I have updated the endpoint in the `MastodonClient`

AnuraagReddy123 commented:
So I just wanted to know if I am going in the right direction
Also in this link https://docs.joinmastodon.org/entities/Announcement/#mentions, it is mentioned that the type of `mentions` is an array `Announcements::Account` and similarly for `statuses`. So should I make nested classes for `Account` and `Status` inside `Announcement` class (Similar to `Poll.kt` in entity? Or can I just use the pre existing Account and Status classes?

andregasser commented:
In this case I would create two new nested data classes. Looking at the documentation, classes are named Announcement::Status and Announcement::Account. I presume they did that to highlight the fact that is not the same as the regular classes we already have.

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/method/AnnouncementMethods.kt:
```diff
@@ -0,0 +1,79 @@
+     * @param withDismissed Boolean If true, response will include announcements dismissed by the user. Defaults to false.
+     * @see <a href="https://docs.joinmastodon.org/methods/announcements/#get">Mastodon API documentation: methods/announcements/#get</a>
+     */
+    fun getAllAnnouncements(
```
Please add the `@JvmOverloads` annotation here. Otherwise, when called from Java, default arguments may not be interpreted correctly or you won’t be able to call this method either with or without the Boolean argument.

```suggestion
    @JvmOverloads
    fun getAllAnnouncements(
```

PattaFeuFeu replied:
I _think_ it’s only necessary for cases where you call `client.performAction` directly.

Right @andregasser?

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/method/AnnouncementMethods.kt:
```diff
@@ -0,0 +1,79 @@
+
+    /**
+     * Undo a react emoji to an announcement.
+     * @param announcementId String The ID of the Announcement in the database.
+     * @param emojiName String Unicode emoji, or the shortcode of a custom emoji.
+     * @see <a href="https://docs.joinmastodon.org/methods/announcements/#delete-reactions">
+     * Mastodon API documentation:methods/announcements/#delete-reactions</a>
+     */
+    fun deleteReaction(
+        announcementId: String,
+        emojiName: String
```
Same suggestion as on the `putReaction` 😊 

```suggestion
     * @param emojiName String Unicode emoji, or the shortcode of a custom emoji.
     * @param announcementId String The ID of the Announcement in the database.
     * @see <a href="https://docs.joinmastodon.org/methods/announcements/#delete-reactions">
     * Mastodon API documentation:methods/announcements/#delete-reactions</a>
     */
    fun removeReactionFromAnnouncement(
        emojiName: String,
        announcementId: String,
```

AnuraagReddy123 replied:
Adding a comma at L72 gives me this error on running `./gradlew check`
```
Unnecessary trailing comma before ")"
```
I will avoid adding a comma there

PattaFeuFeu replied:
Clash between what I’m used to from other projects and the LINT settings for this project. 😵‍💫 
Sure thing 👍

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/method/AnnouncementMethods.kt:
```diff
@@ -0,0 +1,79 @@
+
+    /**
+     * React to an announcement with an emoji.
+     * @param announcementId String The ID of the Announcement in the database.
+     * @param emojiName String Unicode emoji, or the shortcode of a custom emoji.
+     * @see <a href="https://docs.joinmastodon.org/methods/announcements/#put-reactions"> Mastodon API documentation:methods/announcements/#put-reactions</a>
+     */
+    fun putReaction(
+        announcementId: String,
+        emojiName: String
```
I’d suggest to rename this method:

```suggestion
     * @param emojiName String Unicode emoji, or the shortcode of a custom emoji.
     * @param announcementId String The ID of the Announcement in the database.
     * @see <a href="https://docs.joinmastodon.org/methods/announcements/#put-reactions"> Mastodon API documentation:methods/announcements/#put-reactions</a>
     */
    fun addReactionToAnnouncement(
        emojiName: String,
        announcementId: String,
```

If we swap the order to have `emojiName` first and `announcementId` second, it would even lead to a sentence that almost reads natural:

```kotlin
addReactionToAnnouncement(emojiName = 👍, announcementId = "123456")
```

PattaFeuFeu reviewed (commented):
Thank you for your first PR on this project! 🥳

andregasser commented:
Also congrats from my side to your first PR 🙂

AnuraagReddy123 commented:
So what should my next steps be? I have so far:
- Made the entities for Announcement.kt and Reaction.kt
- Made the methods in AnnouncementMethods.kt
- Added the nested classes for Account and Status in Announcement.kt
- Added the api in MastodonClient.kt

PattaFeuFeu commented:
@AnuraagReddy123 Please add an RxJava3 implementation for the `AnnouncementMethods` (`RxAnnouncementMethods` in `bigbone-rx`). You can check out #301 for an example of how this might look. 😊 

For methods that return a value, use `Single.fromCallable { YOUR_METHOD_CALL().execute() }` and for methods that don’t return a value, use `Completable.fromAction { YOUR_METHOD_CALL() }`

Please also add tests at least for the `AnnouncementMethods`. It’s fine if you don’t have them for the Rx variant.

AnuraagReddy123 commented:
The RxCusotmEmojisMethods.kt has the method implemented as
https://github.com/andregasser/bigbone/blob/e40be4915a7904de6f4ed5414760cbca8edb438c/bigbone-rx/src/main/kotlin/social/bigbone/rx/RxCustomEmojisMethods.kt#L22-L24
whereas in RxNotificationMethod.kt it is wrapped inside try-catch block
https://github.com/andregasser/bigbone/blob/e40be4915a7904de6f4ed5414760cbca8edb438c/bigbone-rx/src/main/kotlin/social/bigbone/rx/RxNotificationMethods.kt#L20-L32

Since the `getAllAnnouncements` method is very similar to the `getAllNotifications` method, I was thinking I should implement it how it is given in `getAllNotifications`, which one would you suggest? Also some doubts I had:
- What exactly are the `Single` and `Completable` class 
- What are the differences between the `bigbone-rx` and `bigbone` modules?

PattaFeuFeu commented:
@AnuraagReddy123 Please use the shorter version. See https://github.com/andregasser/bigbone/pull/288 for more info on that one. 😊 

Single and Completable are two of the base types of RxJava. Others are Observable, Maybe, and Flowable (plus Publisher and Subject).

RxJava is the Java implementation of the Reactive Extensions (pattern) which allow stream-based processing of data (very simplified).

see https://github.com/ReactiveX/RxJava

AnuraagReddy123 commented:
> For methods that return a value, use `Single.fromCallable { YOUR_METHOD_CALL().execute() }` and for methods that don’t return a value, use `Completable.fromAction { YOUR_METHOD_CALL() }`

Just a small doubt why are we doing the `.execute()` code part only for the `Single.fromCallable` function and not for `Completable.fromAction` function? 

Also I have implemented the RxAnnouncementMethods.kt file, however I am a bit doubtful if the code is completely correct or not. @PattaFeuFeu can you please check it?

And finally how do I implement the test files? Going through the `NotificationMethodsTest.kt`, I got a general idea of how it can be implemented, but how exactly do I "test" if it is working? Like how do I run the test file?

PattaFeuFeu commented:
> Just a small doubt why are we doing the .execute() code part only for the Single.fromCallable function and not for Completable.fromAction function?

Sure thing, thanks for asking. 😊 
For the methods that return a `MastodonRequest`, we’ll need to call `execute()` on them to actually execute the network call. Otherwise, you would return a `Single<MastodonRequest<TYPE>>` and not a `Single<TYPE>`. When calling `execute()`, internally a pretty similar code as in `client.performRequest` is called.

For the methods that call `performAction` and return `Unit` (i.e.: “nothing”), we don’t need that as `performAction` _already_ executes a network call.

That’s why we need to call `Single.fromCallable { ….execute() }` but `Completable.fromAction { client.performAction(…) }`.

> Also I have implemented the RxAnnouncementMethods.kt file, however I am a bit doubtful if the code is completely correct or not. @PattaFeuFeu can you please check it?

Yep, will do! 😊 

> And finally how do I implement the test files? Going through the NotificationMethodsTest.kt, I got a general idea of how it can be implemented, but how exactly do I "test" if it is working? Like how do I run the test file?

Not sure how it’s done in VSCode. In IntelliJ, you have a “play” button right next to each test and next to the test class:

<img width="239" alt="grafik" src="https://github.com/andregasser/bigbone/assets/9012887/a3827eae-53e9-40d1-8598-01201edfb444">

You can probably google for “VSCode kotlin unit test” or something of that ilk.

But if you call `gradle check`, tests will also be executed.

You can try that out by starting with a test test that will just verify that your tests are ran:

```kotlin
import org.amshove.kluent.shouldBeTrue
import org.junit.jupiter.api.Test

class AnnouncementMethodsTest {

    @Test
    fun `just a test-test`() {
        false.shouldBeTrue()
    }
}
```

That test should of course fail then.

PattaFeuFeu commented on bigbone-rx/src/main/kotlin/social/bigbone/rx/RxAnnouncementMethods.kt:
```diff
@@ -0,0 +1,44 @@
+
+/**
+ * Reactive implementation of [AnnouncementMethods].
+ * Allows access to API methods with endpoints having an "api/vX/announcements" or "api/vX/announcement" prefix.
```
All endpoints seem to use `api/v1/announcements` and none use the singular form, so I’d suggest to change this.

```suggestion
 * Allows access to API methods with endpoints having an "api/vX/announcements" prefix.
```

PattaFeuFeu replied:
@AnuraagReddy123 You marked this as resolved but the part is still there

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/entity/Announcement.kt:
```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Represents an announcement set by an administrator.
+ * @see <a href="https://docs.joinmastodon.org/entities/Announcement/"> Mastodon API Announcement </a>
```
```suggestion
 * @see <a href="https://docs.joinmastodon.org/entities/Announcement/">Mastodon API Announcement</a>
```

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/method/AnnouncementMethods.kt:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * Allows access to API methods with endpoints having an "api/vX/announcements" prefix.
+ * @see <a href="https://docs.joinmastodon.org/methods/notifications/">Mastodon notifications API methods</a>
```
Copy-paste mistake? 😊 

```suggestion
 * @see <a href="https://docs.joinmastodon.org/methods/announcements/">Mastodon announcements API methods</a>
```

AnuraagReddy123 replied:
Oops silly mistake 😅

PattaFeuFeu reviewed (changes requested):
I did not checkout your branch but from a cursory look at the RxAnnouncementMethods, this looks like it’s going in the right direction! 👍

AnuraagReddy123 commented:
Also what should the tests be? I thought I had to just verify each function in AnnouncementMethods.kt but in CustomEmojisMethodsTest.kt
https://github.com/andregasser/bigbone/blob/master/bigbone/src/test/kotlin/social/bigbone/api/method/CustomEmojisMethodsTest.kt
and NotificationMethodsTest.kt
https://github.com/andregasser/bigbone/blob/master/bigbone/src/test/kotlin/social/bigbone/api/method/NotificationMethodsTest.kt
the functions seemed random. So should I randomly write some test functions or is there a list of some test functions somewhere?

AnuraagReddy123 commented:
Sorry if the question is a bit simple, I haven't implemented test cases or used junit before 😅

PattaFeuFeu commented:
They should test the happy path and a few potential error cases. For the methods, there always is a JSON response in the documentation. We copy that into a file in the test assets, reference that file in tests, and then write asserts to ensure that a) the correct method is called, b) the deserialisation works correctly (check for specific properties in the JSON, they should be in the return value of the function if everything went right).

The unhappy paths can be tested by either returning an error in the mock client, or supplying wrong data in case of methods where we have client-side validation (not the case for your PR as far as I have seen).

PattaFeuFeu commented:
@AnuraagReddy123 Did that clarify it a bit? 😊

AnuraagReddy123 commented:
So I tried running the test on Intellij for AnnounementMethodsTest.kt 
`
class AnnouncementMethodsTest {
    @Test
    fun `just a test-test`() {
        true.shouldBeTrue()
    }
}
`
The test case passed
![image](https://github.com/andregasser/bigbone/assets/66117695/0374a078-81b0-46ce-afcf-833037e32a9a)

However for CustomEmojisMethodsTest.kt, for the function at
https://github.com/andregasser/bigbone/blob/c9567d12d28ab534a4a335d85dc3cedb560e368b/bigbone/src/test/kotlin/social/bigbone/api/method/CustomEmojisMethodsTest.kt#L15-L39
I get the following error
```
> Task :bigbone:processTestResources NO-SOURCE
> Task :bigbone:testClasses UP-TO-DATE

java.lang.ExceptionInInitializerError
	at social.bigbone.testtool.MockClient.mock(MockClient.kt:138)
	at social.bigbone.testtool.MockClient.mock$default(MockClient.kt:19)
	at social.bigbone.api.method.CustomEmojisMethodsTest.Given a client returning success, when getting all custom emoji, then expect values of response(CustomEmojisMethodsTest.kt:17)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at java.util.ArrayList.forEach(Unknown Source)
	at java.util.ArrayList.forEach(Unknown Source)
Caused by: java.lang.IllegalStateException: No compatible attachment provider is available
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:627)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:611)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:563)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:540)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.initInstrumentation(JvmMockKAgentFactory.kt:128)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.init(JvmMockKAgentFactory.kt:34)
	at io.mockk.impl.JvmMockKGateway.<init>(JvmMockKGateway.kt:46)
	at io.mockk.impl.JvmMockKGateway.<clinit>(JvmMockKGateway.kt:185)
	... 6 more


No compatible attachment provider is available
java.lang.IllegalStateException: No compatible attachment provider is available
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:627)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:611)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:563)
	at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:540)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.initInstrumentation(JvmMockKAgentFactory.kt:128)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.init(JvmMockKAgentFactory.kt:34)
	at io.mockk.impl.JvmMockKGateway.<init>(JvmMockKGateway.kt:46)
	at io.mockk.impl.JvmMockKGateway.<clinit>(JvmMockKGateway.kt:185)
	at social.bigbone.testtool.MockClient.mock(MockClient.kt:138)
	at social.bigbone.testtool.MockClient.mock$default(MockClient.kt:19)
	at social.bigbone.api.method.CustomEmojisMethodsTest.Given a client returning success, when getting all custom emoji, then expect values of response(CustomEmojisMethodsTest.kt:17)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at java.util.ArrayList.forEach(Unknown Source)
	at java.util.ArrayList.forEach(Unknown Source)
```
Is this the expected behaviour for the test function? I got a similar error in the NotificationMethodsTest.kt file

PattaFeuFeu commented:
No, it is not expected. The tests work both on my local machine as well as in the GitHub Actions builds.

A cursory Google search got me here: https://github.com/raphw/byte-buddy/issues/1148#issuecomment-971640429

Are there setup issues again where your system is using the JRE for some reason?

AnuraagReddy123 commented:
For some reason IntelliJ refuses to work with WSL, so I started to go through the command line documentation from `./gradlew help --task test`. After a lot of trial and error I figured out a way to run a single method through the command line
```
./gradlew :bigbone:test --tests social.bigbone.api.method.AnnouncementMethodsTest."just a test-test"
```
or in other words
```
./gradlew :bigbone:test --tests social.bigbone.api.method.<ClassName>.<MethodName>
```
Through this I was able to run all the tests

PattaFeuFeu commented:
@AnuraagReddy123 If you want to join our Slack, we could do a huddle and I can give you a 101 of testing, IntelliJ, and debugging?

AnuraagReddy123 commented:
Actually that would be really helpful! Thanks a lot!

AnuraagReddy123 commented:
I wrote a test case:
```
fun `Get all unread announcements`() {
      val client = MockClient.mock("announcements.json")
      val announcementMethods = AnnouncementMethods(client)
      val announcements = announcementMethods.getAllAnnouncements(withDismissed = false).execute()

      announcements.size shouldBeEqualTo 1

      verify {
          client.get(
              path = "/api/v1/announcements",
              query = any<Parameters>()
          )
      }
  }
```
According to the Mastodon API for get all announcements [Link](https://docs.joinmastodon.org/methods/announcements/#get), having the withDismissed variable as false should only return those announcements which are not dismissed by the user.

I made this [announcements.json](https://github.com/andregasser/bigbone/files/13218841/announcements.json), where the read variable is false for 2 of the announcements, so calling getAllAnnouncements with withDismissed false should return just 1 announcement, but all 3 announcements are returned

PattaFeuFeu commented:
@AnuraagReddy123 While I get that that would be great, that is not how our mocking setup currently works. 😊 We do not completely emulate the Mastodon API. So what you set in the JSON you give to the `MockClient` is what it will return. There is no filtering for the parameters—that part is done by the actual servers, not by our library.

I commend your effort and agree that this would be a great test _if_ we did the filtering.

But in this case, less is enough. 😊

AnuraagReddy123 commented:
Oh that makes sense. I have added some test cases can you check them @PattaFeuFeu?

PattaFeuFeu commented:
Two small git tips:

1. Ideally, you would create (feature) branches in your fork as well. I noticed that you are working on `AnuraagReddy123:master`, so you are limited to to merge from `andregasser:master` now and cannot merge from your own version of `master`. That in itself is not too bad, but if we merge your Pull Request here, you could not start another PR from your fork. You would first need to get it exactly to the state this repository here is in to start over again as we won’t merge your changes with the exact commits and commits IDs, but will instead “squash” them into one commit, thereby rewriting that part of the git history.
2. Ideally, you would limit commits to one related change only. In [199803d](https://github.com/andregasser/bigbone/pull/309/commits/199803d189491ca4aa392641ad39fcc39d88c1f0) you both added tests _and_ changed a type. Not the case here, but _if_ you now wanted to revert one of the two changes, the work would be more involved then simply reverting the whole commit that only introduced one of the changes

😊

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+        val announcementMethods = AnnouncementMethods(client)
+        val announcements = announcementMethods.getAllAnnouncements(withDismissed = true).execute()
+
+        announcements.size shouldBeEqualTo 3
```
Tip, but optional: You can use
```suggestion
        announcements shouldHaveSize 3
```

(needs `import org.amshove.kluent.shouldHaveSize`)

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+        val client = MockClient.mock("announcements.json")
+        val announcementMethods = AnnouncementMethods(client)
+
+        invoking { announcementMethods.dismissAnnouncement("8") } shouldNotThrow AnyException
```
That’s a bit verbose, but works. Unless explicitly set up that way, unit tests don’t expect exceptions anyway, so the test would fail anyway.

```suggestion
        announcementMethods.dismissAnnouncement("8")
```

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+        val client = MockClient.mock("announcements.json")
+        val announcementMethods = AnnouncementMethods(client)
+
+        invoking {
+            announcementMethods.addReactionToAnnouncement(
+                "😀",
+                "8"
+            )
+        } shouldNotThrow AnyException
```
```suggestion
        announcementMethods.addReactionToAnnouncement(
                "😀",
                "8"
        )
```

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+        val client = MockClient.mock("announcements.json")
+        val announcementMethods = AnnouncementMethods(client)
+
+        invoking {
+            announcementMethods.removeReactionFromAnnouncement(
+                "😀",
+                "8"
+            )
+        } shouldNotThrow AnyException
```
```suggestion
        announcementMethods.removeReactionFromAnnouncement(
                "😀",
                "8"
        )
```

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    fun `Dismiss an announcement and verify that the method was called correctly`() {
+        val client = MockClient.mock("announcements.json")
```
Dismiss, add, and remove methods always return `{}` in their JSON response. I think we should reflect that here. You can add a new file `announcement_dismiss_success.json` with content `{}` and use that here, even if we don’t do anything with the content for now.

AnuraagReddy123 replied:
I have made the changes, but I am not really sure what exactly we are doing here. So in this piece of code:
```
fun `Dismiss an announcement and verify that the method was called correctly`() {
    val client = MockClient.mock("announcement_dismiss_success.json")
    val announcementMethods = AnnouncementMethods(client)

    announcementMethods.dismissAnnouncement("8")

    verify {
        client.performAction(
            endpoint = "/api/v1/announcements/8/dismiss",
            method = MastodonClient.Method.POST
        )
    }
}
```
Just to make sure I understand correctly, 
- We mock the `announcments_dismiss_success.json` file which contains an empty list
- Make the announcementmethods object that takes the client as a part of its constructor
- Dismisses the announcement with announcement id 8
But how would it be able to dismiss the announcement with announcement id 8 when `announcement_dismiss_sucess.json` is empty? Shouldn't it return an error?
- The verify step is clear to me, it just checks whether the client actually called the perform action method with the correct arguements or not

PattaFeuFeu replied:
> But how would it be able to dismiss the announcement with announcement id 8 when announcement_dismiss_sucess.json is empty? Shouldn't it return an error?

Again: The `MockClient` merely returns the content of the JSON file you supply as a successful response. There’s no actual handling of the calls we send to it.

You can check how we set up the `MockClient`. Maybe it makes a bit more sense then. 😊 

For now, as you said, the only thing we really test here is that calling `dismissAnnouncement` then calls the `performAction` function on the `client` with the expected parameters.
We might want to improve that in the future and use a proper OKHTTP MockWebServer. But that’s not currently the case.

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    fun `Add reaction to announcement and verify that the method was called correctly`() {
+        val client = MockClient.mock("announcements.json")
```
`announcements_add_reaction_success.json` with content `{}`

PattaFeuFeu commented on bigbone/src/test/kotlin/social/bigbone/api/method/AnnouncementMethodsTest.kt:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    fun `Remove reaction from announcement and verify that the method was called correctly`() {
+        val client = MockClient.mock("announcements.json")
```
`announcements_remove_reaction_success.json` with content `{}`

PattaFeuFeu reviewed (changes requested):
I really like the direction this is going in! Well done! 😊

AnuraagReddy123 commented:
> Ideally, you would create (feature) branches in your fork as well. I noticed that you are working on `AnuraagReddy123:maste`r, so you are limited to merge from `andregasser:master` now and cannot merge from your own version of master. That in itself is not too bad, but if we merge your Pull Request here, you could not start another PR from your fork. You would first need to get it exactly to the state this repository here is in to start over again as we won’t merge your changes with the exact commits and commits IDs, but will instead “squash” them into one commit, thereby rewriting that part of the git history.

Ok so I should make a new branch like 'AnuraagReddy123:feat-branch', make my changes there, then make a pull request from the feat branch to 'andregasser:master' branch?

> Ideally, you would limit commits to one related change only. In [199803d](https://github.com/andregasser/bigbone/pull/309/commits/199803d189491ca4aa392641ad39fcc39d88c1f0) you both added tests and changed a type. Not the case here, but if you now wanted to revert one of the two changes, the work would be more involved then simply reverting the whole commit that only introduced one of the changes

Yes I will keep that in mind 👍. Currently I follow this workflow in my local machine
- Sync the fork repository with the latest changes![image](https://github.com/andregasser/bigbone/assets/66117695/42ce54d6-06d9-4970-9247-8c6bb00467bf)

- Pull the repository to my laptop
- Make a new feature branch
- Make changes and make multiple commits
- After I am done, I merge this branch to the master branch (in my laptop) using the --squash command (Converts all the commits to a single commit on merging)) (I usually do this to keep the number of commits small, not sure if it is a good practice)
- Push the master branch to Github master branch

PattaFeuFeu commented:
> Ok so I should make a new branch like 'AnuraagReddy123:feat-branch', make my changes there, then make a pull request from the feat branch to 'andregasser:master' branch?

Yep 👍

PattaFeuFeu commented:
> (I usually do this to keep the number of commits small, not sure if it is a good practice)

Depends on the merge type of the project. If the project merges through squash commits like we do here, it doesn’t matter too much. Could be 1 or 100 commits and there would not be a difference on `master` after merging.

For merge commits or rebasing, it would clutter the main branch quite a bit if you had 100 additional commits from one PR, especially if many of them essentially handle the same change.

What has worked _for me_—as long as no one has reviewed yet: Write as many commits as necessary to work well. I often see them as “save points” like in a game, so that I can return to an older solution if something I later worked on didn’t pan out quite as planned.
But then, before pushing for a PR review, I rebase all of the commits onto the source branch (often `main`) and squash commits belonging to the same topic (“add endpoints”, “add entities”, “reformat”) and give them meaningful commit messages.

Regarding the “as long as no one has reviewed yet”: You should try your best to never mess with an already uploaded history. If no one has seen it yet, it could be fine. But as soon as there’s been a review and other people have checked it out, as a general rule of thumb, you shouldn’t rebase or squash existing commits you have already uploaded. That way, you would lose exact references of existing comments as the commits no longer exist. And reviewers would have to freshly check out your changes again instead of just pulling from the remote.
With PRs that’s usually not _that_ big a deal, but you should really not rewrite history on the commonly used base branch (`master`, `main`, `develop`, …) as it’s easy for others to lose their work that way if they don’t know exactly how to use git under those circumstances.

AnuraagReddy123 commented:
So do I have to make any more changes?

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/entity/Announcement.kt:
```diff
@@ -0,0 +1,148 @@
+     * When the announcement will start.
+     */
+    @SerialName("starts_at")
+    val startsAt: String? = null,
```
Let’s update this to match the recent changes implemented in #312:

```suggestion
    @Serializable(with = DateTimeSerializer::class)
    val startsAt: PrecisionDateTime = InvalidPrecisionDateTime.Unavailable,
```

AnuraagReddy123 replied:
Sorry! I guess I forgot to change this again when I was moving from wsl to windows

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/entity/Announcement.kt:
```diff
@@ -0,0 +1,148 @@
+     * When the announcement will end.
+     */
+    @SerialName("ends_at")
+    val endsAt: String? = null,
```
```suggestion
    @Serializable(with = DateTimeSerializer::class)
    val endsAt: PrecisionDateTime = InvalidPrecisionDateTime.Unavailable,
```

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/entity/Announcement.kt:
```diff
@@ -0,0 +1,148 @@
+     * When the announcement was published.
+     */
+    @SerialName("published_at")
+    val publishedAt: String = "",
```
```suggestion
    @Serializable(with = DateTimeSerializer::class)
    val publishedAt: PrecisionDateTime = InvalidPrecisionDateTime.Unavailable,
```

PattaFeuFeu commented on bigbone/src/main/kotlin/social/bigbone/api/entity/Announcement.kt:
```diff
@@ -0,0 +1,148 @@
+     * When the announcement will end.
+     */
+    @SerialName("updated_at")
+    val updatedAt: String? = null,
```
```suggestion
    @Serializable(with = DateTimeSerializer::class)
    val updatedAt: PrecisionDateTime = InvalidPrecisionDateTime.Unavailable,
```

PattaFeuFeu reviewed (changes requested):
I think these are the last changes I can find here that we should change before merging.

AnuraagReddy123 commented:
A small doubt: is the endpoint supposed to be `api/vx/announcements` or `/api/vx/announcments` or are both just the same? Because going through the code, the other methods have the endpoints in the format of `api/vx/...` but in the Mastodon documentation it is written as '/api/vx/...`

PattaFeuFeu commented:
With how our code is written, the version _without_ a leading slash would be better, so `api/vx/…`. But it doesn’t really seem to matter. When running the samples, both versions work for me even if the one _with_ the leading slash produces a double slash in the middle of the URL.

We might fix that in an upcoming PR. But for this PR, as long as you’re using the same variant in all places, it’s fine.

AnuraagReddy123 commented:
All done!

PattaFeuFeu reviewed (approved):
Thank you for sticking with us through all the hurdles your computer threw at you and through the learning curve you had!

AnuraagReddy123 commented:
It was nothing compared to how much I learned about Kotlin and writing good code. Thanks a lot for teaching me so much!
